include/mega/logging.h

```diff
@@ -109,41 +119,223 @@ enum LogLevel {
+
+    template<typename T>
+    typename std::enable_if<std::is_pointer<T>::value && !std::is_same<T, char*>::value>::type
+    logValue(const T value)
+    {
+        char str[20];
+        const auto size = std::sprintf(str, "%p", reinterpret_cast<const void*>(value));
+        copyToBuffer(str, size);
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value && std::is_signed<T>::value
+                            && !std::is_same<T, long>::value && !std::is_same<T, long long>::value>::type
+    logValue(const T value)
+    {
+        char str[20];
+        const auto size = std::sprintf(str, "%d", value);
+        copyToBuffer(str, size);
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value && std::is_signed<T>::value
+                            && std::is_same<T, long>::value && !std::is_same<T, long long>::value>::type
+    logValue(const T value)
+    {
+        char str[20];
+        const auto size = std::sprintf(str, "%ld", value);
+        copyToBuffer(str, size);
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value && std::is_signed<T>::value
+                            && !std::is_same<T, long>::value && std::is_same<T, long long>::value>::type
+    logValue(const T value)
+    {
+        char str[20];
+        const auto size = std::sprintf(str, "%lld", value);
+        copyToBuffer(str, size);
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value && std::is_unsigned<T>::value
+                            && !std::is_same<T, unsigned long>::value && !std::is_same<T, unsigned long long>::value>::type
+    logValue(const T value)
+    {
+        char str[20];
+        const auto size = std::sprintf(str, "%u", value);
+        copyToBuffer(str, size);
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value && std::is_unsigned<T>::value
+                            && std::is_same<T, unsigned long>::value && !std::is_same<T, unsigned long long>::value>::type
+    logValue(const T value)
+    {
+        char str[20];
+        const auto size = std::sprintf(str, "%lu", value);
+        copyToBuffer(str, size);
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value && std::is_unsigned<T>::value
+                            && !std::is_same<T, unsigned long>::value && std::is_same<T, unsigned long long>::value>::type
+    logValue(const T value)
+    {
+        char str[20];
+        const auto size = std::sprintf(str, "%llu", value);
+        copyToBuffer(str, size);
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_floating_point<T>::value>::type
+    logValue(const T value)
+    {
+        char str[20];
+        const auto size = std::sprintf(str, "%f", value);
+        copyToBuffer(str, size);
+    }
+
+    void logValue(const char* value)
+    {
+        copyToBuffer(value, std::strlen(value));
```

Comment:
I'm wondering if we should consider implementing this by looping until `\0`, to avoid one extra loop (strlen) before the reading loop that copyToBuffer entails. We should gain a little bit in performance and this should be (one of ) the most used cases. Shouldn't it?

Reply:
Yes, good point. Will change accordingly.

Reply:
Actually, I am not convinced anymore: `std::strlen` and `std::copy` will be highly optimized and possibly vectorized. Also, the check `if (mBufferIt == mBuffer.end() - 1)` only has to be done after the bulk copy but if we copy each element then this check has to be done every time.
What do you think?

Btw, this is how I'd implement what you suggested:
```cpp
    void logValue(const char* value)
    {
        while (*value)
        {
            *mBufferIt = *value;
            ++mBufferIt;
            if (mBufferIt == mBuffer.end() - 1)
            {
                outputBuffer();
            }
            ++value;
        }
    }
```

Reply:
I think to get a definite answer on this we'd have to measure across different string sizes. However, the real big fish to fry will be in MEGAsync itself where logging seems quite inefficient.